File: 4.depth-first-search/근우/5.validate-binary-search-tree.py
## https://leetcode.com/problems/validate-binary-search-tree/

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
import sys
from collections import deque
from typing import Optional

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(lower_bound, upper_bound,node):
            if not node:
                return True
            if lower_bound<node.val<upper_bound:
                return all([dfs(lower_bound,node.val,node.left),dfs(node.val,upper_bound,node.right)])    
            return False
        return dfs(-sys.maxsize,sys.maxsize,root) 
    

File: 4.depth-first-search/동규/5.validate-binary-search-tree.js
// https://leetcode.com/problems/validate-binary-search-tree/.

function isValidBST(root) {
  let preVal = -Infinity;
  let stack = [];

  while (root !== null || stack.length !== 0) {
    while (root !== null) {
      stack.push(root);
      root = root.left;
    }
    root = stack.pop();
    if (root.val <= preVal) return false;
    preVal = root.val;
    root = root.right;
  }
  return true;
}
